utils/src/main/java/org/akvo/flow/utils/XmlFormParser.kt

```diff
@@ -322,8 +323,227 @@ class XmlFormParser @Inject constructor(private val helper: FileHelper) {
             groups = groups)
     }
 
-    private fun getIntAttribute(parser: XmlPullParser, attributeName: String) =
-        parser.getAttributeValue(null, attributeName)?.toInt() ?: -1
+    fun parseXmlQuestions(inputStream: InputStream): HashMap<Int, MutableList<Question>> {
+        val parserFactory: XmlPullParserFactory
+        var currentQuestion: Question? = null
+        var currentOptions = mutableListOf<Option>()
+        var currentOption: Option? = null
+        var currentLevels = mutableListOf<Level>()
+        var currentLevel: Level? = null
+        var currentAltText: AltText? = null
+        var currentHelp: QuestionHelp? = null
+        var lastText: String? = null
+        var groupOrder = 0
+        var currentQuestions = mutableListOf<Question>()
+        val map = HashMap<Int, MutableList<Question>>()
+        try {
+            parserFactory = XmlPullParserFactory.newInstance()
+            val parser = parserFactory.newPullParser()
+            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
+            parser.setInput(inputStream, null)
+            var eventType = parser.eventType
+            while (eventType != XmlPullParser.END_DOCUMENT) {
+
+                when (eventType) {
+                    /**
+                     * Beginning of a tag for example <question>
+                     */
+                    XmlPullParser.START_TAG -> {
```

Comment:
Just a thought based on an example of something I saw, wouldnt it be easier if the sections here were broken down into separate functions per type of element? e.g. whats done here..

https://developer.android.com/training/basics/network-ops/xml#parse

Reply:
it's not that easy because we have some sections that are common to questions, options , but I can take a look at the end of the work done of how it can be improved